packages/workbox-window/src/Workbox.ts

```diff
@@ -502,13 +502,20 @@ class Workbox extends WorkboxEventTarget {
    */
   private readonly _onControllerChange = (originalEvent: Event) => {
     const sw = this._sw;
-    if (sw === navigator.serviceWorker.controller) {
-      this.dispatchEvent(new WorkboxEvent('controlling', {
-        sw,
-        originalEvent,
-        isUpdate: this._isUpdate,
-      }));
+    const isExternal = sw !== navigator.serviceWorker.controller;
 
+    // Unconditionally dispatch the controlling event, with isExternal set
+    // to distinguish between controller changes due to the initial registration
+    // vs. an update-check or other tab's registration.
+    // See https://github.com/GoogleChrome/workbox/issues/2786
+    this.dispatchEvent(new WorkboxEvent('controlling', {
+      isExternal,
+      originalEvent,
+      sw,
+      isUpdate: this._isUpdate,
+    }));
+
+    if (!isExternal) {
```

Comment:
I decided that the `controlling` promise would _not_ resolve (if it hasn't already been resolved) if the new controller is an "external" service worker, as there is no means to distinguish between expected/external in that case.